Firestore/core/src/model/object_value.cc

```diff
@@ -0,0 +1,329 @@
+                  absl::string_view segment) const {
+    return nanopb::MakeStringView(entry.key) < segment;
+  }
+  bool operator()(absl::string_view segment,
+                  const google_firestore_v1_MapValue_FieldsEntry& entry) const {
+    return segment < nanopb::MakeStringView(entry.key);
+  }
+};
+
+/**
+ * Finds an entry by key in the provided map value. Returns `nullptr` if the
+ * entry does not exist.
+ */
+google_firestore_v1_MapValue_FieldsEntry* FindEntry(
+    const google_firestore_v1_Value& value, absl::string_view segment) {
+  if (value.which_value_type != google_firestore_v1_Value_map_value_tag) {
+    return nullptr;
+  }
+  const google_firestore_v1_MapValue& map_value = value.map_value;
+
+  // MapValues in iOS are always stored in sorted order.
+  auto found = std::equal_range(map_value.fields,
+                                map_value.fields + map_value.fields_count,
+                                segment, MapEntryKeyCompare());
+
+  if (found.first == found.second) {
+    return nullptr;
+  }
+
+  return found.first;
+}
+
+size_t CalculateSizeOfUnion(
+    google_firestore_v1_MapValue* map_value,
+    const std::map<std::string, google_firestore_v1_Value>& upserts,
+    const std::set<std::string>& deletes) {
+  // Compute the size of the map after applying all mutations. The final size is
+  // the number of existing entries, plus the number of new entries
+  // minus the number of deleted entries.
+  return upserts.size() +
+         std::count_if(
+             map_value->fields, map_value->fields + map_value->fields_count,
+             [&](const google_firestore_v1_MapValue_FieldsEntry& entry) {
+               std::string field = MakeString(entry.key);
+               // Don't count if entry is deleted or if it is a replacement
+               // rather than an insert.
+               return deletes.find(field) == deletes.end() &&
+                      upserts.find(field) == upserts.end();
+             });
+}
+
+/**
+ * Modifies `parent_map` by adding, replacing or deleting the specified
+ * entries.
+ */
+void ApplyChanges(
+    google_firestore_v1_MapValue* parent,
+    const std::map<std::string, google_firestore_v1_Value>& upserts,
+    const std::set<std::string>& deletes) {
+  // TODO(mrschmidt): Consider using `absl::btree_map` and `absl::btree_set` for
+  // potentially better performance.
+  auto source_count = parent->fields_count;
+  auto* source_fields = parent->fields;
+
+  size_t target_count = CalculateSizeOfUnion(parent, upserts, deletes);
+  auto* target_fields =
+      MakeArray<google_firestore_v1_MapValue_FieldsEntry>(target_count);
+
+  auto delete_it = deletes.begin();
+  auto upsert_it = upserts.begin();
+
+  // Merge the existing data with the deletes and updates
+  for (pb_size_t source_index = 0, target_index = 0;
+       target_index < target_count;) {
+    if (source_index < source_count) {
+      std::string key = MakeString(source_fields[source_index].key);
```

Comment:
Note: Using a string_view here causes an ASAN failure as the key may get freed using FreeNanopbMessage in the loop.

Reply:
Hmm, but `key` doesn't seem to be used after `FreeNanopbMessage`? AFAICS, the memory can only be freed _after_ a comparison and it causes the loop to continue.